tests.py

```diff
@@ -0,0 +1,212 @@
+
+    _generate_tests(TestIPlot, box_test, 'box', options)
+
+
+def area_plot_input_argument_tests():
+    options = {
+        'fill': [True],
+        'opacity': [1],
+        'kind': ['area']
+    }
+
+    def area_test(self, **kwargs):
+        self._iplot(self.df, **kwargs)
+
+    _generate_tests(TestIPlot, area_test, 'area', options)
+
+
+def scatter_plot_input_argument_tests():
+    options = {
+        'x': ['x'],
+        'y': ['y'],
+        'mode': ['markers'],
+        'symbol': ['dot'],
+        'colors': [['orange', 'teal']],
+        'size': [10]
+    }
+
+    def scatter_test(self, **kwargs):
+        self._iplot(self.df, **kwargs)
+
+    _generate_tests(TestIPlot, scatter_test, 'scatter', options)
+
+
+def bubble_chart_argument_tests():
+    options = {
+        'x': ['x'], 'y': ['y'], 'size': ['c']
+    }
+
+    def bubble_test(self, **kwargs):
+        self._iplot(self.df, **kwargs)
+
+    _generate_tests(TestIPlot, bubble_test, 'bubble', options)
+
+
+def subplot_input_argument_tests():
+    options = {
+        'shape': [(3, 1)],
+        'shared_xaxes': [True],
+        'vertical_spacing': [0.02],
+        'fill': [True],
+        'subplot_titles': [True],
+        'legend': [False]
+    }
+
+    def subplot_test(self, **kwargs):
+        self._iplot(self.df, subplots=True, **kwargs)
+
+    _generate_tests(TestIPlot, subplot_test, 'subplots', options)
+
+
+def shape_input_argument_tests():
+    df = cf.datagen.lines(3, columns=['a', 'b', 'c'])
+    options = {
+        'hline': [
+            [2, 4],
+            [dict(y=-1, color='blue', width=3),
+             dict(y=1, color='pink', dash='dash')]],
+        'vline': [['2015-02-10']],
+        'hspan': [[(-1, 1), (2, 5)]],
+        'vspan': [{
+            'x0': '2015-02-15', 'x1': '2015-03-15',
+            'color': 'teal', 'fill': True, 'opacity': .4}]
+    }
+
+    def shape_tests(self, **kwargs):
+        self._iplot(df, **kwargs)
+
+    _generate_tests(TestIPlot, shape_tests, 'shape', options)
+
+# test generators
+
+
+def _generate_tests(test_class, test_func, test_name, options):
+    from itertools import chain, combinations, product
+
+    def powerset(iterable):
+        "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
+        s = list(iterable)
+        return chain.from_iterable(combinations(s, r)
+                                   for r in range(len(s) + 1))
+
+    key_value_tuple = {}
+    for option, values in options.iteritems():
+        key_value_tuple[option] = [(option, i) for i in values]
+
+    for option_groups in powerset(key_value_tuple.values()):
+        for input_kwargs in product(*option_groups):
+            kwargs = {i[0]: i[1] for i in input_kwargs}
+            setattr(
+                test_class,
+                'test_{}_{}'.format(test_name, '__'.join([
+                    '_'.join([str(s) for s in i])
+                    for i in kwargs.items()])),
+                _generate_test(test_func, **kwargs))
```

Comment:
this adds the generated tests to the TestClass above. And since we're adding attributes that start with the name "test", they'll automatically get run